p2p/tools/local_network.py

```diff
@@ -0,0 +1,378 @@
+    reader = asyncio.StreamReader()
+
+    transport = AddressedTransport(address, reader)
+    protocol = asyncio.StreamReaderProtocol(reader)
+
+    writer = asyncio.StreamWriter(
+        transport=transport,
+        protocol=protocol,
+        reader=reader,
+        loop=asyncio.get_event_loop(),
+    )
+    return reader, writer
+
+
+def direct_pipe() -> ReaderWriterPair:
+    reader = asyncio.StreamReader()
+
+    transport = MemoryTransport(reader)
+    protocol = asyncio.StreamReaderProtocol(reader)
+
+    writer = asyncio.StreamWriter(
+        transport=transport,
+        protocol=protocol,
+        reader=reader,
+        loop=asyncio.get_event_loop(),
+    )
+    return reader, writer
+
+
+async def _connect_streams(
+        reader: asyncio.StreamReader,
+        writer: asyncio.StreamWriter,
+        queue: asyncio.Queue,
+        token: CancelToken) -> None:
+    try:
+        while not token.triggered:
+            if reader.at_eof():
+                break
+
+            try:
+                size = queue.get_nowait()
+            except asyncio.QueueEmpty:
+                await asyncio.sleep(0)
+                continue
+            data = await wait_with_token(reader.readexactly(size), token=token)
+            writer.write(data)
+            queue.task_done()
+            await wait_with_token(writer.drain(), token=token)
+    except OperationCancelled:
+        pass
+    finally:
+        writer.write_eof()
+
+    if reader.at_eof():
+        reader.feed_eof()
+
+
+class Router(BaseService):
+    hosts: Dict[str, 'Host']
+    networks = Dict[str, 'Network']
+    connections = Dict[CancelToken, asyncio.Task]
+
+    def __init__(self):
+        super().__init__()
+        self.hosts = {}
+        self.networks = {}
+        self.connections = {}
+
+    #
+    # Service API
+    #
+    async def _run(self) -> None:
+        while not self.cancel_token.triggered:
+            await asyncio.sleep(0.02)
+
+    async def _cleanup(self) -> None:
+        # all of the cancel tokens *should* be triggered already so we just
+        # wait for the networking processes to complete.
+        if self.connections:
+            await asyncio.wait(
+                self.connections.values(),
+                timeout=2,
+                return_when=asyncio.ALL_COMPLETED
+            )
+
+    #
+    # Connections API
+    #
+    def get_host(self, host: str) -> 'Host':
+        if host not in self.hosts:
+            self.hosts[host] = Host(host, self)
+        return self.hosts[host]
+
+    def get_network(self, name: str) -> 'Network':
+        if name not in self.networks:
+            self.networks[name] = Network(name, self)
+        return self.networks[name]
+
+    def get_connected_readers(self, address: Address) -> ReaderWriterPair:
+        external_reader, internal_writer = direct_pipe()
+        internal_reader, external_writer = addressed_pipe(address)
+
+        token = CancelToken(str(address)).chain(self.cancel_token)
```

Comment:
I guess this is the biggest pain point for replacing with `asyncio.Event`. If you want to pull the network mocking into its own distro, maybe the canceltoken stuff has to get pulled out too so we can depend on it in both p2p and mock_network?